### PR TITLE
[WIP] Add a Dockerfile.

### PR DESCRIPTION
**Pull request by d33tah**
Not sure yet what this is for - I was just testing how difficult it would be
to build the thing under Dockerfile. It was trivial. Let me know if you'd
like me to drop it or adapt it to any specific use case.

### CODE REVIEW

**Comment by TheJJ** [comment]
What would be the value of that dockerfile in the repo? I'm not familiar with docker's usage.
Would it work to place the file in a subdirectory (e.g. `etc/` or `dist`)?

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+FROM ubuntu:18.04
+
+RUN apt-get update && apt-get -y install cmake cython3 libeigen3-dev libepoxy-dev libfontconfig1-dev libfreetype6-dev libharfbuzz-dev libogg-dev libopus-dev libopusfile-dev libpng-dev libsdl2-dev libsdl2-image-dev python3-dev python3-jinja2 python3-numpy python3-pil python3-pip python3-pygments qml-module-qtquick-controls qtdeclarative5-dev && apt-get clean
+
```
Comment (stefaang) [inline Dockerfile]: Why would you apt update 3 times? Better to combine all in a single command.
Improve readability by putting each dependency on a separate line with a trailing backslash (makes it easier to maintain).

**Comment by mic-e** [comment]
I think it would make sense to test this in `kevin`.

File: Dockerfile (hunk 1 above)
Comment (d33tah) [reply Dockerfile]: @stefaang will do :)

**Comment by d33tah** [comment]
@TheJJ right now it's just an experiment - I wanted to see if it was my setup or Ubuntu in general that had problems with building and this Dockerfile shows that builds work fine under Ubuntu. Docker provides a reproducible environment for this kind of tests.

The file can't be put into `etc` because Dockerfiles can't reference parent directories and we need files from them. This is why I put it into the root directory of the repo.

Stay tuned for an update - as the title says, it's currently an early WIP.

**Comment by d33tah** [comment]
It seems that I am getting CI errors unrelated to my patch.

**Comment by zuntrax** [comment]
I'll look into those errors.

**Comment by simonsan** [comment]
Merged the recent master into this branch. Checks now passing.

@d33tah any updates regarding the WIP status of this PR?

**Comment by simonsan** [comment]
One month has passed now since we were waiting for feedback. I'm closing this PR and move it to the zombies, so we can get back to it at a later stage/when we think about doing anything with docker.

**Comment by kyleVsteger** [comment]
@d33tah I just tried to run build this dockerfile as is and `FROM ubuntu:20.04` without any luck.
20.04 seems to get further along. Any insight? I'd like to experiment with X11/Xquartz forwarding from an ubuntu container to make this easier to run on mac.
```
docker build -t openage:ubuntu .
Sending build context to Docker daemon  21.74MB
Step 1/6 : FROM ubuntu:20.04
 ---> f643c72bc252
Step 2/6 : ENV DEBIAN_FRONTEND="noninteractive"
 ---> Using cache
 ---> 3f42cb10af71
Step 3/6 : RUN apt-get update && apt-get install     cmake     cython3     flex     git     libeigen3-dev     libepoxy-dev     libfontconfig1-dev     libfreetype6-dev     libharfbuzz-dev     libogg-dev     libopus-dev     libopusfile-dev     libpng-dev     libsdl2-dev     libsdl2-image-dev     make     python3-dev     python3-jinja2     python3-numpy     python3-pil     python3-pip     python3-pygments     qml-module-qtquick-controls     qtdeclarative5-dev     -y && apt-get clean
 ---> Using cache
 ---> 7cc6995b30a0
Step 4/6 : WORKDIR /tmp
 ---> Using cache
 ---> de4b66deabb0
Step 5/6 : ADD . .
 ---> Using cache
 ---> 1aa19a9b948a
Step 6/6 : RUN ./configure --download-nyan && make
 ---> Running in b1863cc33bba


 ___  ______ _______ _______ ___
|  _)/ _____|_______|_______|_  |
| | ( (____  _____      _     | |    ___  ____  _____ ____  _____  ____ _____
| |  \____ \|  ___)    | |    | |   / _ \|  _ \| ___ |  _ \(____ |/ _  | ___ |
| |_ _____) ) |        | |   _| |  | |_| | |_| | ____| | | / ___ ( (_| | ____|
|___|______/|_|        |_|  (___|   \___/|  __/|_____)_| |_\_____|\___ |_____)
                                         |_|                     (_____|

Welcome to the SFT technologies computer-aided openage build system!

You have chosen, or been chosen, to attempt the daring task of building openage.
If you have installed all the dependencies that are conveniently listed in
[doc/building.md], this _might_ just work!

If it doesn't, consider reporting the issue: https://github.com/SFTtech/openage
Or ask for help:
  * Matrix: #sfttech:matrix.org
  * IRC:    #sfttech on freenode.net

-- The CXX compiler identification is GNU 9.3.0
-- Check for working CXX compiler: /usr/bin/g++
-- Check for working CXX compiler: /usr/bin/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test HAVE_GSPLIT_DWARF_SUPPORT
-- Performing Test HAVE_GSPLIT_DWARF_SUPPORT - Success
-- Performing Test HAVE_REQUIRED_CXX17_SUPPORT
-- Performing Test HAVE_REQUIRED_CXX17_SUPPORT - Success
-- Found Python3: /usr/bin/python3.8 (found suitable exact version "3.8.5") found components: Interpreter Development NumPy 
-- Found Python: /usr/bin/python3.8 (Required is at least version "3.6") 
-- Found Cython: /usr/bin/python3.8 -m cython (found suitable version "0.29.14", minimum required is "0.25") 
-- Checking python3 module PIL.Image - Success
-- Checking python3 module PIL.ImageDraw - Success
-- Checking python3 module numpy - Success
-- Checking python3 module pygments - Success
-- Checking python3 module jinja2 - Success
-- Set PROJECT_VERSION from git.
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
-- The C compiler identification is GNU 9.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
CMake Warning at buildsystem/doxygen.cmake:50 (message):
  doxygen couldn't be found, you won't be able to generate docs
Call Stack (most recent call first):
  CMakeLists.txt:219 (doxygen_configure)


-- Found Freetype: /usr/lib/x86_64-linux-gnu/libfreetype.so (found version "2.10.1") 
-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.11") 
-- Found PNG: /usr/lib/x86_64-linux-gnu/libpng.so (found version "1.6.37") 
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found SDL2: /usr/include/SDL2  
-- Found SDL2Image: /usr/lib/x86_64-linux-gnu/libSDL2_image.so  
-- Found Opusfile: /usr/lib/libopusfile.so  
-- Found Epoxy: /usr/lib/x86_64-linux-gnu/libepoxy.so  
-- Found HarfBuzz: /usr/lib/x86_64-linux-gnu/libharfbuzz.so (found suitable version "2.6.4", minimum required is "1.0.0") 
-- Could NOT find GCCBacktrace (missing: GCCBacktrace_LIBRARIES GCCBacktrace_INCLUDE_DIRS) 
-- Could NOT find nyan (missing: nyan_DIR)
-- Downloading nyan as submodule project...
Cloning into 'source'...
Note: switching to 'origin/master'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at a899eb2 Merge pull request #81 from heinezen/spec-rework


       ▄▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▄
      █░░▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒░░░█
      █░▒▒▒▒▒▒▒▒▒▒▄▀▀▄▒▒▒░░█▄▀▀▄
 ▄▄   █░▒▒▒▒▒▒▒▒▒▒█▓▓▓▀▄▄▄▄▀▓▓▓█      ▄ ▄▄   ▄   ▄   ▄▄▄   ▄ ▄▄
█▓▓█▄▄█░▒▒▒▒▒▒▒▒▒▄▀▓▓▓▓▓▓▓▓▓▓▓▓▀▄     █▀  █  ▀▄ ▄▀  ▀   █  █▀  █
 ▀▄▄▓▓█░▒▒▒▒▒▒▒▒▒█▓▓▓▄█▓▓▓▄▓▄█▓▓█     █   █   █▄█   ▄▀▀▀█  █   █
    ▀▀█░▒▒▒▒▒▒▒▒▒█▓▒▒▓▄▓▓▄▓▓▄▓▒▒█     █   █   ▀█    ▀▄▄▀█  █   █
     ▄█░░▒▒▒▒▒▒▒▒▒▀▄▓▓▀▀▀▀▀▀▀▓▄▀              ▄▀
   ▄▀▓▀█▄▄▄▄▄▄▄▄▄▄▄▄██████▀█▀▀               ▀▀
   █▄▄▀ █▄▄▀       █▄▄▀ ▀▄▄█


For information about building, look at [doc/building.md].

If you encounter problems you can't fix yourself,
report problems at https://github.com/SFTtech/nyan

Contact: #sfttech:matrix.org

-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
-- Found FLEX: /usr/bin/flex (found suitable version "2.6.4", minimum required is "2.6") 

          project | nyan
          version | 0.1
         compiler | GNU 9.3.0
       build type | Debug
         cxxflags |  -fdiagnostics-color=auto -Wall -Wextra -pedantic
 build type flags | -O0 -ggdb3
        build dir | /tmp/.bin/g++-debug-Oauto-sanitize-none
   install prefix | /usr/local

-- nyan processed successfully!
-- optional dependency is unavailable: backtrace
-- Could NOT find GPerfTools (missing: GPERFTOOLS_LIBRARIES GPERFTOOLS_INCLUDE_DIR) 
-- optional dependency is unavailable: gperftools-profiler
-- Found Inotify: /usr/include/x86_64-linux-gnu  
-- Could NOT find Curses (missing: CURSES_LIBRARY CURSES_INCLUDE_PATH) 
-- optional dependency is unavailable: ncurses
-- Found OpenGL: /usr/lib/x86_64-linux-gnu/libOpenGL.so   
-- Found Vulkan: /usr/lib/x86_64-linux-gnu/libvulkan.so  

cython module
	run                                                 [embedded interpreter] [noinstall]
	openage.cython_check
	openage.cabextract.lzxd
	openage.cabextract.cabchecksum                      [standalone]
	openage.convert.service.export.interface.visgrep
-- Found Ogg: /usr/lib/x86_64-linux-gnu/libogg.so  
	openage.convert.service.export.opus.opusenc
	openage.convert.service.export.png.png_create
	openage.convert.value_object.read.media.slp
	openage.convert.value_object.read.media.smp
	openage.convert.value_object.read.media.smx
	openage.cppinterface.exctranslate
	openage.cppinterface.exctranslate_tests
	openage.cppinterface.pyobject
	openage.cppinterface.setup_checker
	openage.cvar.cvar
	openage.event.demo
	openage.game.main_cpp
	openage.log.log_cpp
	openage.util.filelike.cpp
	openage.util.fslike.cpp
	openage.renderer.renderer_cpp
	openage.renderer.tests
	openage.testing.cpp_testing
	openage.testing.misc_cpp
	openage.versions.versions
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/tmp/.bin/g++-debug-Oauto-sanitize-none/openage/__main__.py", line 139, in <module>
    sys.exit(main())
  File "/tmp/.bin/g++-debug-Oauto-sanitize-none/openage/__main__.py", line 89, in main
    from .convert.main import init_subparser
  File "/tmp/.bin/g++-debug-Oauto-sanitize-none/openage/convert/main.py", line 17, in <module>
    from .service.init.version_detect import create_version_objects
  File "/tmp/.bin/g++-debug-Oauto-sanitize-none/openage/convert/service/init/version_detect.py", line 8, in <module>
    import toml
ModuleNotFoundError: No module named 'toml'
CMake Error at buildsystem/codegen.cmake:32 (message):
  failed to get target list from codegen invocation
Call Stack (most recent call first):
  buildsystem/python.cmake:494 (codegen_run)
  CMakeLists.txt:257 (python_finalize)


-- Configuring incomplete, errors occurred!
See also "/tmp/.bin/g++-debug-Oauto-sanitize-none/CMakeFiles/CMakeOutput.log".
See also "/tmp/.bin/g++-debug-Oauto-sanitize-none/CMakeFiles/CMakeError.log".
./configure is a convenience script:
it creates the build directory,  symlinks it,
and invokes cmake for an out-of-source build.

Nobody is stopping you from skipping ./configure and our Makefile,
and using CMake directly (e.g. when packaging, or using an IDE).
For your convenience, ./configure even prints the direct CMake invocation!

         build_type | Debug
       cxx_compiler | g++
          cxx_flags | 
   exe_linker_flags | 
     install_prefix | /usr/local
module_linker_flags | 
shared_linker_flags | 

config options:

          backtrace | if_available
gperftools-profiler | if_available
gperftools-tcmalloc | False
            inotify | if_available
            ncurses | if_available
             opengl | if_available
             vulkan | if_available

bindir:
/tmp/.bin/g++-debug-Oauto-sanitize-none/

invocation:
cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_COMPILER=g++ -DCMAKE_CXX_FLAGS='' -DCMAKE_EXE_LINKER_FLAGS='' -DCMAKE_INSTALL_PREFIX=/usr/local -DCMAKE_MODULE_LINKER_FLAGS='' -DCMAKE_SHARED_LINKER_FLAGS='' -DDOWNLOAD_NYAN=YES -DCXX_OPTIMIZATION_LEVEL=auto -DCXX_SANITIZE_FATAL=False -DCXX_SANITIZE_MODE=none -DWANT_BACKTRACE=if_available -DWANT_GPERFTOOLS_PROFILER=if_available -DWANT_GPERFTOOLS_TCMALLOC=False -DWANT_INOTIFY=if_available -DWANT_NCURSES=if_available -DWANT_OPENGL=if_available -DWANT_VULKAN=if_available -- /tmp

(now running cmake:)
```

**Comment by kyleVsteger** [comment]
I've started a new thread in regards to this here https://github.com/SFTtech/openage/pull/1357

**Pull request closed without merging**